[CORE-14941] `ct/l1`: remove `extent`s in `replace_objects()` below start offset

WillemKauf opened the pull request:
It is not hard to imagine a concurrent race between a `replace_objects()` and a `set_start_offset()` update to the `metastore` (e.g. a race between compaction of a log and a prefix truncation).
Both outcomes of the race are at the present time handled correctly:

1. If `replace_objects()` lands first, `set_start_offset()` will simply discard whatever extents are in the `metastore` below the new start offset.
2. If `set_start_offset()` lands first, the `replace_objects()` request will _likely_ be rejected (it may still be accepted if the offset update does not remove any existing `extent`s).

Case 2 is currently handled pessimistically. We do not need to reject a `replace_objects()` command just because some of the objects in the update are below the new start offset- it does not affect the validity of the objects above the new start offset.

Handle this case better by removing extents in an update that are fully below a partition's start offset before checking the validity of the update. This is performed in a new free function `remove_extents_below_start_offset()`.

If _all_ new extents for _all_ partitions in the update are removed an error is returned to the caller.



## Backports Required

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.3.x
- [ ] v25.2.x
- [ ] v25.1.x

## Release Notes

* none

WillemKauf committed:
`ct/l1`: soften requirements in `replace_objects()` below start offset

It is not hard to imagine a concurrent race between a `replace_objects()`
and a `set_start_offset()` update to the `metastore` (e.g. a race between
compaction of a log and a prefix truncation).
Both outcomes of the race are at the present time handled correctly:

1. If `replace_objects()` lands first, `set_start_offset()` will simply
   discard whatever extents are in the `metastore` below the new start offset.
2. If `set_start_offset()` lands first, the `replace_objects()` request will
   _likely_ be rejected (it may still be accepted if the offset update
   does not remove any existing `extent`s).

Case 2 is currently handled pessimistically. We do not need to reject
a `replace_objects()` command just because some of the objects in the update
are below the new start offset- it does not affect the validity of the objects
above the new start offset.

Handle this case better by tweaking the logic in `get_range()` to ignore
extent alignment requirements for replacements below the log start offset,
and then remove the un-needed extents after they have been added in `apply()`.

We allow for misalignment between the start offset and the first extent in
the partition, since we need to support partial truncations- this also means
we don't need replacement extents to be aligned to existing extents below the
start offset. We will still need to validate the alignment of the range's
last offset to be sure the update is valid.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--ca895a8d-2035-4280-b2b2-07c318d81628-->
<!--TEST RESULTS-->
<!--13797b84-98d5-4921-a8e3-4dc31f4a0a2a-->
<!--TEST RESULTS-->
<!--1fd799fa-6fd9-4394-9299-84e770ac4f14-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77531">build#77531</a></summary>


| test_class                                         | test_method                                        | test_arguments                     | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                                 |
|:---------------------------------------------------|:---------------------------------------------------|:-----------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| NodesDecommissioningTest                           | test_decommissioning_rebalancing_node              | {"shutdown_decommissioned": false} | integration | https://buildkite.com/redpanda/redpanda/builds/77531#019aff52-23d3-430f-a065-4b964c8f3b0b | FLAKY         | 8/11     | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0964, p0=0.2499, reject_threshold=0.0100. adj_baseline=0.2621, p1=0.4895, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_rebalancing_node |
| WriteCachingFailureInjectionE2ETest                | test_crash_all                                     | {"use_transactions": false}        | integration | https://buildkite.com/redpanda/redpanda/builds/77531#019aff52-23d3-430f-a065-4b964c8f3b0b | FLAKY         | 9/11     | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0836, p0=0.5825, reject_threshold=0.0100. adj_baseline=0.2305, p1=0.2907, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all             |
| src/v/cloud_storage/tests/segment_meta_cstore_test | src/v/cloud_storage/tests/segment_meta_cstore_test |                                    | unit        | https://buildkite.com/redpanda/redpanda/builds/77531#019aff17-3dfa-4696-b1cf-3ab655f6d26b | FAIL          | 0/1      |                                                                                                                                                                                                                                                                                                                                                                                                     |                                                                                                                                                                              |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77542">build#77542</a></summary>


| test_class                    | test_method                           | test_arguments                                        | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                                 |
|:------------------------------|:--------------------------------------|:------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests | test_topic_delete                     | {"source_cluster_spec": {"cluster_type": "redpanda"}} | integration | https://buildkite.com/redpanda/redpanda/builds/77542#019affe8-ab49-47ca-9754-384f49f26733 | FLAKY         | 23/31    | Test FAILS after retries.[Significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-B%3A-Test-fails-with-an-improbably-higher-rate-than-baseline-(run-fails))(baseline=0.0823, p0=0.0096, reject_threshold=0.0100)                                                                    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_topic_delete                |
| NodesDecommissioningTest      | test_decommissioning_rebalancing_node | {"shutdown_decommissioned": true}                     | integration | https://buildkite.com/redpanda/redpanda/builds/77542#019affec-1d52-4b7f-83d5-4389fe336dad | FLAKY         | 8/11     | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0961, p0=0.2487, reject_threshold=0.0100. adj_baseline=0.2614, p1=0.4917, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_rebalancing_node |
| ScalingUpTest                 | test_moves_with_local_retention       | {"use_topic_property": true}                          | integration | https://buildkite.com/redpanda/redpanda/builds/77542#019affec-1d54-4399-bc5e-e23ad3dc9a59 | FLAKY         | 29/31    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0264, p0=0.5513, reject_threshold=0.0100. adj_baseline=0.0770, p1=0.3164, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ScalingUpTest&test_method=test_moves_with_local_retention                  |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77585">build#77585</a></summary>


| test_class      | test_method          | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason   | test_history   |
|:----------------|:---------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------|:---------------|
| MasterTestSuite | test_partial_upload2 |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/77585#019b0409-eb38-4831-afda-3448688d53ad | FAIL          | 0/1      |          |                |
</details>
<!--/body-->

andrwng commented on src/v/cloud_topics/level_one/metastore/state_update.cc:
```diff
@@ -374,6 +430,12 @@ replace_objects_update::can_apply(const state& state) {
     }
 
+    auto remove_res = remove_extents_below_start_offset<false>(
+      new_extents_by_tp, state);
```
I think it will be more correct if we only remove extents after applying the replacement. This seems like it could end up failing an otherwise valid replacement operation.

State is `[0, 100]`
Replacement is is `[0, 30][31, 60][61, 100]`
Start offset at time of replacement is 75, so the replacement is `[61, 100]`, which doesn't align with `[0, 100]`.

But we should probably instead accept the replacement and then truncate to `[61, 100]`. Could add this as a test case?

WillemKauf replied:
Really good catch, thank you.

I solved this a slightly different way, by considering the "effective" start offset to be the minimum between the reported log offset and the first extent's base offset (if present) before attempting to truncate extents out of the update.

Also added your test case, thanks!

LMKWYT.

andrwng replied:
Just an initial read, i think the template and "effective start" make the code more confusing and difficult to follow, in an area of the code where it is critical to be understandable and correct. Do you have qualms against just removing the extents after apply?

WillemKauf replied:
> the template 

I don't have a strong opinion on making the template parameter a regular function parameter, assuming that the compiler does a good enough job of const folding to come up with the same branch removal for the `if` logging condition.

> Do you have qualms against just removing the extents after apply?

The original scenario I was thinking of with this fix was a full truncation like follows:

```
Initial state: [0, 30][31, 60][61, 100]
(Compaction job begins)
set_start_offset{61} -> [61,100]
(Compaction job completes)
compact_objects([0, 30][31, 60][61, 100]) -> rejected due to alignment with state [61,100]
```

which means we can't just remove the extents after apply, as far I can see.

LMK if I'm missing something or if you can see a clearer solution here.

andrwng replied:
Got it.

My gripe isn't really around the template. It's more around there being special cases where the code operates in different modes under different conditions -- I find that different modes tend to make the code harder to reason about. My hope was to reduce this into a not special case by removing on apply, but yea maybe that isn't feasible. Will give it some more thought.

andrwng replied:
How about this instead: when calling `get_range()`, allow the request base offset to be below the partition extent start as long as the request last offset matches; and then after applying, remove the extents that fall below the start offset.

I like this in that it's more intuitive for me to think about this as 1) accepting the full update, and then 2) truncating to something valid. I'm wary of truncating the update and expecting the application of the truncated update to result in valid state (I don't have a specific case in mind, but it seems brittle)

WillemKauf replied:
This would also work, though semantically it feels like we're ultimately doing the same thing, just in a reverse order (in my solution, we truncate then apply, in yours, apply then truncate)

It feels _less_ brittle to me to truncate first, build up the contiguous intervals, perform the validation checks, and then be sure the truncated update is still valid.

> expecting the application of the truncated update to result in valid state (I don't have a specific case in mind, but it seems brittle)

Isn't this an argument for truncating & validating before applying? I also can't draw up a specific case where this would lead to an invalid update, but if I could, we would have caught it in the validation step.

andrwng replied:
>Isn't this an argument for truncating & validating before applying? I

No, I'm wary that `expecting the [truncated update] to result in valid state` will not hold true.

>It feels less brittle to me to truncate first, build up the contiguous intervals, perform the validation checks, and then be sure the truncated update is still valid.

IMO the metastore doesn't really have the right to truncate the input request in this way.

A replacement in my mind is an atomic unit to be applied. Removing extents via set_start_offset is also something that is an atomic unit to be applied. It therefore feels safe and natural to accept extents completely, and then remove those that need to be removed, as this is a sequence of two atomic operations. Provided the extents are still contiguous, expanding the bounds of the extents seems like a more natural price to pay than truncating the input.

By truncating the input, it feels like we're violating a contract that we're applying our updates atomically, and that's what's bothering me.

In any case, let's take another example:

```
state: [11, 20][21, 30]
start offset: 15
replacement: [0, 15][16, 30]
```

If we're to truncate the input, the effective start offset is 11, and we don't end up truncating the input. The contiguous range is [0, 30], so we call `get_range(0, 30)` and end up with a validation error, when this is actually valid and can result in [16, 30].

WillemKauf replied:
> state: [11, 20][21, 30]
> start offset: 15
> replacement: [0, 15][16, 30]
> 
> If we're to truncate the input, the effective start offset is 11, and we don't end up truncating the input. The contiguous range is [0, 30], so we call get_range(0, 30) and end up with a validation error, when this is actually valid and can result in [16, 30].

Fair enough, I'll look at 

> when calling get_range(), allow the request base offset to be below the partition extent start as long as the request last offset matches; and then after applying, remove the extents that fall below the start offset

soon

WillemKauf replied:
Pushed. LMKWYT.

WillemKauf replied:
~~Actually hmm. This doesn't address [the case I was first concerned about.](https://github.com/redpanda-data/redpanda/pull/28897#discussion_r2600194523)~~

~~Revisiting.~~

andrwng replied:
Awesome, thank you for making the change. Left one (maybe functional?) suggestion that I think will make it easier to reason about

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/77542-->
<!--b6b88f57-fd28-421f-a96f-9e84736ada6d-->
<!--title-->
## Retry command for [Build#77542](https://buildkite.com/redpanda/redpanda/builds/77542)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
skip-redpanda-build
skip-units
skip-rebase
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkingReplicationTests.test_topic_delete@{"source_cluster_spec":{"cluster_type":"redpanda"}}
```
<!--/body-->

dotnwat reviewed (commented):
like the optimization. will defer to andrew for approval

> If _all_ new extents for _all_ partitions in the update are removed an error
is returned to the caller.

does the caller do anything different in the _all removed_ case vs the _some removed_ case?

WillemKauf commented:
> does the caller do anything different in the all removed case vs the some removed case?

No, _some_ removed will apply the valid updates for the other partitions.

dotnwat commented:
> > does the caller do anything different in the all removed case vs the some removed case?
> 
> No, _some_ removed will apply the valid updates for the other partitions.

Right, so then follow up question is why is an error return specifically in the case that all extents are removed? isn't that a generalization of the _some_ case? (not suggesting we do anything different than is currently being done, just trying to understand the distinction).

WillemKauf committed:
`ct/l1`: add `remove_extents_below_start_offset_for_tp()`

A helper function which will be used to de-duplicate repeated
code in a future commit, but currently replaces an existing
inline use case.

WillemKauf commented:
Sorry for force pushes, ready for re-review now @andrwng

andrwng commented on src/v/cloud_topics/level_one/metastore/state_update.cc:
```diff
@@ -25,12 +25,26 @@ struct extent_range {
+        } else {
+            return std::nullopt;
+        }
+    }
```
I think it'll be easier to reason about if we don't bother with the lower_bound when base < log_start_offset and instead return the start of the extents in that case, something like:

```
auto base_it = extents.begin();
if (base >= log_start_offset) {
    base_it = std::ranges::lower_bound(
      extents, base, std::less<>{}, &extent::base_offset);
    if (base_it == extents.end() || base_it->base_offset != base) {
        return std::nullopt;
    }
}
```

WillemKauf replied:
Done
